### ISSUE TITLE
False positive used-before-assignment (E0601) for PEP 695 type-parameter bound forward-referencing a later-defined class

### ISSUE THREAD

**Issue by kdkavanagh-jt** [PyCQA/pylint#11115]
### Bug description
A PEP 695 type-parameter bound that forward-references a class defined later in the same module is reported as used-before-assignment (E0601), even though it is valid at runtime. PEP 695 bounds are lazily evaluated (stored in a code object and only evaluated on TypeVar.__bound__ access), so referring to a name defined later in the module is legal.

This looks like the same class of false positive as #9815 (used-before-assignment for type aliases), but the fix there (PR #10488) only covered the type X = ... alias-value scope, not type-parameter bounds on classes/functions.

```python
class Container[T: Item]:  # E0601 reported on `Item`
    value: T


class Item:
    pass
```

### Configuration

```ini

```

### Command used

```shell
pylint repro.py
```

### Pylint output

```python
************* Module repro
repro.py:1:19: E0601: Using variable 'Item' before assignment (used-before-assignment)
```

### Expected behavior

No error

### Pylint version

```shell
pylint 4.0.5
astroid 4.0.4
Python 3.12.6 (main, Oct  9 2024, 11:19:43) [GCC 13.2.0]
```

### Additional dependencies

```python

```

**Comment by FredStober** [comment]
I hope the proposed fix will also solve the following false positive (seems related...) along with it:

```
from typing import TYPE_CHECKING, Annotated
from botocore.client import BaseClient

if TYPE_CHECKING:
      from mypy_boto3_sns.client import SNSClient
else:
      type SNSClient = Annotated[BaseClient, 'SNSClient']

def f() -> SNSClient:
      """doc."""
```

E0601: Using variable 'SNSClient' before assignment (used-before-assignment)
